pg_search/src/bootstrap/create_bm25.rs

```diff
@@ -358,3 +362,79 @@ fn validate_checksum(index: PgRelation) -> Result<SetOfIterator<'static, String>
 
 #[pg_extern(sql = "")]
 fn create_bm25_jsonb() {}
+
+#[allow(clippy::type_complexity)]
+#[pg_extern]
+fn storage_info(
+    index: PgRelation,
+) -> TableIterator<'static, (name!(block, i64), name!(max_offset, i32))> {
+    let segment_components =
+        LinkedItemList::<SegmentMetaEntry>::open(index.oid(), SEGMENT_METAS_START, false);
+    let bman = segment_components.bman();
+    let mut blockno = segment_components.get_start_blockno();
+    let mut data = vec![];
+
+    while blockno != pg_sys::InvalidBlockNumber {
+        let buffer = bman.get_buffer(blockno);
+        let page = buffer.page();
+        let max_offset = page.max_offset_number();
+        data.push((blockno as i64, max_offset as i32));
+        blockno = page.next_blockno();
+    }
+
+    TableIterator::new(data)
+}
+
+#[allow(clippy::type_complexity)]
+#[pg_extern]
+fn page_info(
+    index: PgRelation,
+    blockno: i64,
+) -> anyhow::Result<
+    TableIterator<
+        'static,
+        (
+            name!(offsetno, i32),
+            name!(size, i32),
+            name!(visible, bool),
+            name!(recyclable, bool),
+            name!(contents, JsonB),
+        ),
+    >,
+> {
+    let segment_components =
+        LinkedItemList::<SegmentMetaEntry>::open(index.oid(), SEGMENT_METAS_START, false);
+    let bman = segment_components.bman();
+    let buffer = bman.get_buffer(blockno as pg_sys::BlockNumber);
+    let page = buffer.page();
+    let max_offset = page.max_offset_number();
+
+    if max_offset == pg_sys::InvalidOffsetNumber {
+        return Ok(TableIterator::new(vec![]));
+    }
+
+    let mut data = vec![];
+    let snapshot = unsafe { pg_sys::GetActiveSnapshot() };
+    let heap_oid = unsafe { pg_sys::IndexGetRelation(index.oid(), false) };
+    let heap_relation = unsafe { pg_sys::RelationIdGetRelation(heap_oid) };
+
+    for offsetno in pg_sys::FirstOffsetNumber..=max_offset {
+        unsafe {
+            if let Some(pg_item) = page.read_pg_item(offsetno) {
```

Comment:
This should use `Page::read_item<T>()`

Reply:
I did this to get the size of the item. Otherwise I'd have to convert the entry into a `PgItem` which feels like double the work.

Reply:
Then change the read_item() function to also return the size.

the problem with putting this on Buffer is that it’s a thing under the purview of Page.